chain/migrate.go

```diff
@@ -0,0 +1,159 @@
+	types.DecodeSliceCast[oldSiafundElement](d, &ts.SiafundInputs)
+	types.DecodeSliceCast[oldFileContractElement](d, &ts.RevisedFileContracts)
+	types.DecodeSliceFn(d, &ts.StorageProofs, func(d *types.Decoder) (sp consensus.V1StorageProofSupplement) {
+		(*oldFileContractElement)(&sp.FileContract).DecodeFrom(d)
+		return
+	})
+}
+
+type oldBlockSupplement consensus.V1BlockSupplement
+
+func (oldBlockSupplement) Cast() (bs consensus.V1BlockSupplement) { return }
+
+func (bs *oldBlockSupplement) DecodeFrom(d *types.Decoder) {
+	types.DecodeSliceCast[oldTransactionSupplement](d, &bs.Transactions)
+	types.DecodeSliceCast[oldFileContractElement](d, &bs.ExpiringFileContracts)
+}
+
+type oldSupplementedBlock supplementedBlock
+
+func (sb *oldSupplementedBlock) DecodeFrom(d *types.Decoder) {
+	if v := d.ReadUint8(); v != 2 {
+		d.SetErr(fmt.Errorf("incompatible version (%d)", v))
+	}
+	var b types.Block
+	(*types.V2Block)(&b).DecodeFrom(d)
+	sb.Block = &b
+	sb.Header = b.Header()
+	types.DecodePtrCast[oldBlockSupplement](d, &sb.Supplement)
+}
+
+type versionedState consensus.State
+
+func (vs *versionedState) DecodeFrom(d *types.Decoder) {
+	if v := d.ReadUint8(); v != 2 {
+		d.SetErr(fmt.Errorf("incompatible version (%d)", v))
+	}
+	(*consensus.State)(vs).DecodeFrom(d)
+}
+
+// MigrateDB upgrades the database to the latest version.
+func MigrateDB(db DB, n *consensus.Network) error {
+	if db.Bucket(bVersion) == nil {
+		return nil // nothing to migrate
+	}
+	dbs := &DBStore{
+		db: db,
+		n:  n,
+	}
+	var err error
+	rewrite := func(bucket []byte, key []byte, from types.DecoderFrom, to types.EncoderTo) {
+		if err != nil {
+			return
+		}
+		b := dbs.bucket(bucket)
+		val := b.getRaw(key)
+		if val == nil {
+			return
+		}
+		d := types.NewBufDecoder(val)
+		from.DecodeFrom(d)
+		if d.Err() != nil {
+			err = d.Err()
+			return
+		}
+		b.put(key, to)
+		if dbs.shouldFlush() {
+			dbs.Flush()
+		}
+	}
+
+	version := dbs.bucket(bVersion).getRaw(bVersion)
+	if len(version) != 1 {
+		return errors.New("invalid version")
+	}
+	switch version[0] {
+	case 1:
+		var sb supplementedBlock
+		for _, key := range db.BucketKeys(bBlocks) {
+			rewrite(bBlocks, key, (*oldSupplementedBlock)(&sb), &sb)
+		}
+		var cs consensus.State
+		for _, key := range db.BucketKeys(bStates) {
+			rewrite(bStates, key, (*versionedState)(&cs), &cs)
+		}
+		var sce types.SiacoinElement
+		for _, key := range db.BucketKeys(bSiacoinElements) {
+			rewrite(bSiacoinElements, key, (*oldSiacoinElement)(&sce), &sce)
+		}
+		var sfe types.SiafundElement
+		for _, key := range db.BucketKeys(bSiafundElements) {
+			rewrite(bSiafundElements, key, (*oldSiafundElement)(&sfe), &sfe)
+		}
+		var fce types.FileContractElement
+		for _, key := range db.BucketKeys(bFileContractElements) {
+			if len(key) == 32 {
+				rewrite(bFileContractElements, key, (*oldFileContractElement)(&fce), &fce)
+			}
+		}
+		if err != nil {
+			return err
+		}
+		dbs.bucket(bVersion).putRaw(bVersion, []byte{2})
```

Comment:
Doesn't this break when it gets interrupted? Since it's not atomic?

Reply:
yeah, not really sure how best to handle that. One thing I tried was "if a value decodes successfully into the new type, then it's already been converted so you can skip it." But that turns out to have false positives: that is, there are encodings that can be decoded into either the old type or the new type without causing an error.

we can control when flushes occur, so maybe the best we can do is store a "progress" entry every time we flush, e.g. `1 5912` would indicate that we've updated all blocks and 5912 states. That implies that `BucketKeys` must return the keys in the same order every time, though.

Reply:
Best to make it atomic instead of trying to hack around atomicity.

Reply:
That might cause problems though. IIRC, bbolt has a max transaction size.

Reply:
right, a single transaction isn't gonna be viable if you're re-encoding every block.

tbh... this kinda is an argument for sticking a version byte in front of every entry 😬

Reply:
Or a saner database 😛